File: Design-Patterns-Using-CSharp/Mediator/MediatorPattern.cs
using System;
using System.Collections.Generic;

interface IChatRoom
{
    void AddParticipant(IParticipant participant);
    void Broadcast(string message, string origin);
    void SendMessage(string message, string to);
}

class ChatRoom : IChatRoom
{
    private Dictionary<string, IParticipant> participants = new Dictionary<string, IParticipant>();

    public void AddParticipant(IParticipant participant)
    {
        participants[participant.GetName()] = participant;
    }

    public void Broadcast(string message, string origin)
    {
        Console.WriteLine($"ChatRoom broadcast Message : {message}");
        foreach (var participant in participants.Values)
        {
            if (participant.GetName() != origin)
            {
                participant.Receive(message);
            }
        }
    }

    public void SendMessage(string message, string to)
    {
        participants[to].Receive(message);
    }
}

interface IParticipant
{
    string GetName();
    void Broadcast(string message);
    void Send(string message, string to);
    void Receive(string message);
}

class Participant : IParticipant
{
    private string name;
    private IChatRoom chatRoom;

    public Participant(string name, IChatRoom chatRoom)
    {
        this.name = name;
        this.chatRoom = chatRoom;
        chatRoom.AddParticipant(this);
    }

    public string GetName()
    {
        return name;
    }

    public void Broadcast(string message)
    {
        Console.WriteLine($"{name} broadcast Message : {message}");
        chatRoom.Broadcast(message, name);
    }

    public void Send(string message, string to)
    {
        Console.WriteLine($"{name} sent Message : {message}");
        chatRoom.SendMessage(message, to);
    }

    public void Receive(string message)
    {
        Console.WriteLine($"{name} received Message : {message}");
    }
}

// Client code

class MediatorPattern
{
    static void Main(string[] args)
    {
        ChatRoom chatRoom = new ChatRoom();
        Participant james = new Participant("James", chatRoom);
        Participant michael = new Participant("Michael", chatRoom);
        Participant robert = new Participant("Robert", chatRoom);

        michael.Send("Good Morning.", "James");
        james.Broadcast("Hello, World!");
    }
}

/*
Michael sent Message : Good Morning.
James received Message : Good Morning.
James broadcast Message : Hello, World!
ChatRoom broadcast Message : Hello, World!
Michael received Message : Hello, World!
Robert received Message : Hello, World!
*/

File: Design-Patterns-Using-Ruby/Mediator/MediatorPattern.rb
class ChatRoom
    def initialize
        @participants = {}
    end

    def add_participant(participant)
        @participants[participant.name] = participant
    end

    def broadcast(message, origin)
        puts "ChatRoom broadcast Message : #{message}"
        @participants.values.each do |p|
            p.receive(message) unless p.name == origin
        end
    end

    def send_message(message, to)
        @participants[to].receive(message)
    end
end

class Participant
    attr_reader :name, :chat_room

    def initialize(name, chat_room)
        @name = name
        @chat_room = chat_room
        chat_room.add_participant(self)
    end

    def broadcast(message)
        puts "#{name} broadcast Message : #{message}"
        chat_room.broadcast(message, name)
    end

    def send_message(message, to)
        puts "#{name} sent Message : #{message}"
        chat_room.send_message(message, to)
    end

    def receive(message)
        puts "#{name} received Message : #{message}"
    end
end

# Client code
chat_room = ChatRoom.new
james = Participant.new("James", chat_room)
michael = Participant.new("Michael", chat_room)
robert = Participant.new("Robert", chat_room)

michael.send_message("Good Morning.", "James")
james.broadcast("Hello, World!")

=begin 
Michael sent Message : Good Morning.
James received Message : Good Morning.
James broadcast Message : Hello, World!
ChatRoom broadcast Message : Hello, World!
Michael received Message : Hello, World!
Robert received Message : Hello, World!
=end